File: DungeonDebt/Assets/Scripts/Core/GameManager.cs
using System;
using UnityEngine;

public class GameManager : MonoBehaviour
{
    [SerializeField] private RunManager _runManager;
    [SerializeField] private ShopManager _shopManager;
    [SerializeField] private PayrollManager _payrollManager;
    [SerializeField] private EncounterManager _encounterManager;

    private GameState _currentState = GameState.MainMenu;

    public event Action<GameState> OnStateChanged;

    public GameState CurrentState
    {
        get { return _currentState; }
    }

    public RunState CurrentRunState
    {
        get
        {
            if (_runManager == null)
            {
                return null;
            }

            return _runManager.CurrentRunState;
        }
    }

    public RunManager RunManager
    {
        get { return _runManager; }
    }

    public ShopManager ShopManager
    {
        get { return _shopManager; }
    }

    public PayrollManager PayrollManager
    {
        get { return _payrollManager; }
    }

    public EncounterManager EncounterManager
    {
        get { return _encounterManager; }
    }

    private void Awake()
    {
        EnsureManagers();
    }

    public void Initialize(RunManager runManager)
    {
        _runManager = runManager;
        EnsureManagers();
    }

    public void StartRun()
    {
        ChangeState(GameState.StartRun);
        ChangeState(GameState.Scout);
    }

    public void ContinueFromScout()
    {
        ChangeState(GameState.Shop);
    }

    public void ContinueFromShop()
    {
        ChangeState(GameState.Formation);
    }

    public void ContinueFromFormation()
    {
        ChangeState(GameState.Payroll);
    }

    public void SelectPayrollAction(PayrollActionId? actionId)
    {
        RunState runState = CurrentRunState;
        if (runState == null)
        {
            return;
        }

        runState.SelectedPayrollAction = actionId;
    }

    public void ContinueFromPayroll()
    {
        EnsureManagers();
        RunState runState = CurrentRunState;
        if (runState != null && runState.SelectedPayrollAction.HasValue && _payrollManager != null)
        {
            _payrollManager.Apply(runState, runState.SelectedPayrollAction.Value);
        }

        ChangeState(GameState.Combat);
    }

    public void ContinueAfterReward()
    {
        EnsureManagers();
        if (_runManager == null)
        {
            return;
        }

        GameState nextState = _runManager.EvaluateNextState();
        if (nextState == GameState.Shop)
        {
            _runManager.AdvanceRound();
            nextState = GameState.Scout;
        }

        RunState runState = CurrentRunState;
        if (runState != null)
        {
            runState.SelectedPayrollAction = null;
        }

        ChangeState(nextState);
    }

    public void ChangeState(GameState nextState)
    {
        EnsureManagers();
        _currentState = nextState;

        if (_currentState == GameState.StartRun && _runManager != null)
        {
            _runManager.InitializeRun();
        }

        if (_currentState == GameState.Scout && _encounterManager != null)
        {
            RunState runState = CurrentRunState;
            int round = runState != null ? runState.Round : 1;
            _encounterManager.LoadEncounter(round);
        }

        if (_currentState == GameState.Shop && _shopManager != null)
        {
            _shopManager.GenerateOffers();
        }

        if (OnStateChanged != null)
        {
            OnStateChanged(_currentState);
        }
    }

    private void EnsureManagers()
    {
        if (_runManager == null)
        {
            _runManager = GetComponent<RunManager>();
            if (_runManager == null)
            {
                _runManager = gameObject.AddComponent<RunManager>();
            }
        }

        if (_shopManager == null)
        {
            _shopManager = GetComponent<ShopManager>();
            if (_shopManager == null)
            {
                _shopManager = gameObject.AddComponent<ShopManager>();
            }
        }

        _shopManager.Initialize(_runManager);

        if (_payrollManager == null)
        {
            _payrollManager = GetComponent<PayrollManager>();
            if (_payrollManager == null)
            {
                _payrollManager = gameObject.AddComponent<PayrollManager>();
            }
        }

        _runManager.Initialize(_payrollManager);

        if (_encounterManager == null)
        {
            _encounterManager = GetComponent<EncounterManager>();
            if (_encounterManager == null)
            {
                _encounterManager = gameObject.AddComponent<EncounterManager>();
            }
        }

        _encounterManager.Initialize(_runManager);
    }
}

File: web/src/core/GameManager.js
// Ported from DungeonDebt/Assets/Scripts/Core/GameManager.cs
// Owns GameState and all state transitions. In Unity, managers were
// MonoBehaviours wired via EnsureManagers + the scene; here GameManager
// constructs and wires them directly. Combat resolution lived in the combat
// view in Unity; resolveCombat() exposes it so the UI (and headless tests) can
// drive a fight and feed the result back through RunManager.

import { GameState } from "./GameState.js";
import { GameRules } from "./GameRules.js";
import { RunManager } from "../run/RunManager.js";
import { ShopManager } from "../run/ShopManager.js";
import { PayrollManager } from "../run/PayrollManager.js";
import { EncounterManager } from "../run/EncounterManager.js";
import { RivalManager } from "../run/RivalManager.js";
import { CombatManager } from "../combat/CombatManager.js";
import { BalanceRunLogger } from "../run/BalanceRunLogger.js";

export class GameManager {
  constructor() {
    this._currentState = GameState.MainMenu;
    this._pendingDifficulty = GameRules.DefaultDifficultyLevel;
    this._stateListeners = [];
    this._highestBeatenDifficulty = -1;

    this._payrollManager = new PayrollManager();
    this._rivalManager = new RivalManager();
    this._runManager = new RunManager(this._payrollManager, this._rivalManager);
    this._shopManager = new ShopManager(this._runManager);
    this._encounterManager = new EncounterManager(this._runManager);
    this._combatManager = new CombatManager();
  }

  get currentState() { return this._currentState; }
  get currentRunState() { return this._runManager ? this._runManager.currentRunState : null; }
  get runManager() { return this._runManager; }
  get shopManager() { return this._shopManager; }
  get payrollManager() { return this._payrollManager; }
  get encounterManager() { return this._encounterManager; }
  get rivalManager() { return this._rivalManager; }
  get highestBeatenDifficulty() { return this._highestBeatenDifficulty; }

  isDifficultyLocked(difficulty) {
    if (!difficulty || !difficulty.isImplemented) return true;
    if (difficulty.level === 0) return false;
    return difficulty.level > this._highestBeatenDifficulty + 1;
  }

  // Replaces the C# `event Action<GameState> OnStateChanged`.
  onStateChanged(listener) {
    this._stateListeners.push(listener);
    return () => {
      const idx = this._stateListeners.indexOf(listener);
      if (idx !== -1) this._stateListeners.splice(idx, 1);
    };
  }

  startRun(difficultyLevel = this._pendingDifficulty) {
    this._pendingDifficulty = difficultyLevel;
    this.changeState(GameState.StartRun);
    this.changeState(GameState.Scout);
  }

  returnToMainMenu() {
    this.changeState(GameState.MainMenu);
  }

  continueFromScout() { this.changeState(GameState.Shop); }
  continueFromShop() { this.changeState(GameState.Formation); }
  continueFromFormation() { this.changeState(GameState.Payroll); }

  selectPayrollAction(actionId) {
    const runState = this.currentRunState;
    if (!runState) return;
    runState.selectedPayrollAction = actionId;
  }

  continueFromPayroll() {
    const runState = this.currentRunState;
    if (runState && runState.selectedPayrollAction !== null && runState.selectedPayrollAction !== undefined && this._payrollManager) {
      this._payrollManager.apply(runState, runState.selectedPayrollAction);
    }
    this.changeState(GameState.Combat);
  }

  // Resolve the current encounter and fold the result into the run. Returns the
  // CombatResult so the UI can replay its log. Mirrors what the Unity combat
  // view did before transitioning to the Reward state.
  resolveCombat() {
    const runState = this.currentRunState;
    if (!runState) return null;
    const encounter = runState.currentEncounter;
    const result = this._combatManager.startCombat(runState, encounter);
    this._runManager.applyPostCombatResult(result, encounter);
    return result;
  }

  continueAfterReward() {
    if (!this._runManager) return;

    const nextState = this._runManager.evaluateNextState();
    const runState = this.currentRunState;
    BalanceRunLogger.logRound(runState, nextState);
    if (runState) runState.selectedPayrollAction = null;

    if (this._runManager.tryPreparePendingRelicReward(nextState)) {
      this.changeState(GameState.RelicReward);
      return;
    }

    this.changeState(nextState);
  }

  continueAfterRelicReward(relicId) {
    if (!this._runManager) return;
    const nextState = this._runManager.selectPendingRelic(relicId);
    this.changeState(nextState);
  }

  continueToNextAct() {
    if (!this._runManager) return;
    this._runManager.advanceToNextAct();
    this.changeState(GameState.Scout);
  }

  continueFromRivalUpdate() {
    if (!this._runManager) return;
    this._runManager.advanceRound();
    this.changeState(GameState.Scout);
  }

  changeState(nextState) {
    this._currentState = nextState;

    if (this._currentState === GameState.StartRun && this._runManager) {
      this._runManager.initializeRun(this._pendingDifficulty);
    }

    if (this._currentState === GameState.Scout && this._encounterManager) {
      const runState = this.currentRunState;
      const round = runState ? runState.round : 1;
      this._encounterManager.loadEncounter(round);
    }

    if (this._currentState === GameState.Shop && this._shopManager) {
      this._shopManager.generateOffers();
    }

    if (this._currentState === GameState.RivalUpdate && this._rivalManager) {
      this._rivalManager.advanceRivals(this.currentRunState);
    }

    if (this._currentState === GameState.Victory && this._runManager) {
      const run = this._runManager.currentRunState;
      if (run && run.selectedDifficulty !== null && run.selectedDifficulty > this._highestBeatenDifficulty) {
        this._highestBeatenDifficulty = run.selectedDifficulty;
      }
    }

    for (const listener of this._stateListeners) {
      listener(this._currentState);
    }
  }
}
